aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -2,11 +2,47 @@
 #include <ATen/native/ReduceOpsUtils.h>
 #include <ATen/native/TensorIterator.h>
 #include <ATen/Parallel.h>
+#include <unordered_map>
+#include <map>
 
 namespace at { namespace native {
 
 namespace {
 
+#define DEFINE_REDUCE_FUNCTOR(FUNCTOR_NAME, OPERATOR) class FUNCTOR_NAME { \
+public:                                                             \
+ FUNCTOR_NAME() {};                                                 \
+ template <typename T>                                              \
+ void operator() (T* self_data, const T* src_data) {                \
+   *self_data OPERATOR *src_data;                                   \
+ };};                                                               \
+ FUNCTOR_NAME FUNCTOR_NAME##_func;
```

Comment:
You can eliminate the macro by using yet another template, and then calling on func(self_data, src_data).
We flip the order for gather. So we can call things like ScatterReduce and GatherReduce..

Reply:
Could you check again? There's something remaining like adding `const` to some of the strings but I'll do that with the addition of the scalar scatter reduction function.